jina/peapods/runtimes/grpc/__init__.py

```diff
@@ -0,0 +1,280 @@
+        self.logger.debug('Teardown GRPCDataRuntime')
+
+        self._data_request_handler.close()
+        self._loop.stop()
+        self._loop.close()
+
+        super().teardown()
+
+    def _close_grpclet(self):
+        self._loop.create_task(self._grpclet.close())
+        self._grpclet_task.cancel()
+
+    @staticmethod
+    def get_control_address(**kwargs):
+        """
+        Does return None, exists for keeping interface compatible with ZEDRuntime
+
+        :param kwargs: extra keyword arguments
+        :returns: None
+        """
+        return None
+
+    @staticmethod
+    def is_ready(ctrl_address: str, **kwargs) -> bool:
+        """
+        Check if status is ready.
+
+        :param ctrl_address: the address where the control message needs to be sent
+        :param kwargs: extra keyword arguments
+
+        :return: True if status is ready else False.
+        """
+
+        try:
+            response = Grpclet.send_ctrl_msg(ctrl_address, 'STATUS')
+        except RpcError:
+            return False
+
+        return True
+
+    @staticmethod
+    def activate(
+        **kwargs,
+    ):
+        """
+        Does nothing
+        :param kwargs: extra keyword arguments
+        """
+        pass
+
+    @staticmethod
+    def cancel(
+        control_address: str,
+        **kwargs,
+    ):
+        """
+        Cancel this runtime by sending a TERMINATE control message
+
+        :param control_address: the address where the control message needs to be sent
+        :param kwargs: extra keyword arguments
+        """
+        Grpclet.send_ctrl_msg(control_address, 'TERMINATE')
+
+    @staticmethod
+    def wait_for_ready_or_shutdown(
+        timeout: Optional[float],
+        ctrl_address: str,
+        shutdown_event: Union['multiprocessing.Event', 'threading.Event'],
+        **kwargs,
+    ):
+        """
+        Check if the runtime has successfully started
+
+        :param timeout: The time to wait before readiness or failure is determined
+        :param ctrl_address: the address where the control message needs to be sent
+        :param shutdown_event: the multiprocessing event to detect if the process failed
+        :param kwargs: extra keyword arguments
+        :return: True if is ready or it needs to be shutdown
+        """
+        timeout_ns = 1000000000 * timeout if timeout else None
+        now = time.time_ns()
+        while timeout_ns is None or time.time_ns() - now < timeout_ns:
+            if shutdown_event.is_set() or GRPCDataRuntime.is_ready(ctrl_address):
+                return True
+            time.sleep(0.1)
+
+        return False
+
+    async def _callback(self, msg: 'Message') -> None:
+        try:
+            msg = self._post_hook(self._handle(self._pre_hook(msg)))
+            asyncio.create_task(self._grpclet.send_message(msg))
+        except RuntimeTerminated:
+            # this is the proper way to end when a terminate signal is sent
+            self._close_grpclet()
```

Comment:
await all `_close_grpclet` calls?

```suggestion
            await self._close_grpclet()
```